modules/misc/listenbrainz.c

```diff
@@ -0,0 +1,585 @@
+            {
+                    .on_update = OnTimerUpdate,
+                    .on_discontinuity = OnTimerStopped,
+            };
+
+    vlc_playlist_t *playlist = p_sys->playlist = vlc_intf_GetMainPlaylist (p_intf);
+    vlc_player_t *player = vlc_playlist_GetPlayer (playlist);
+
+    vlc_playlist_Lock (playlist);
+    p_sys->playlist_listener = vlc_playlist_AddListener (playlist, &playlist_cbs, p_intf, false);
+    if ( !p_sys->playlist_listener )
+    {
+        vlc_playlist_Unlock (playlist);
+        goto error;
+    }
+
+    p_sys->player_listener = vlc_player_AddListener (player, &player_cbs, p_intf);
+    vlc_playlist_Unlock (playlist);
+    if ( !p_sys->player_listener )
+        goto error;
+
+    p_sys->timer_listener = vlc_player_AddTimer (player, VLC_TICK_FROM_SEC (1), &timer_cbs, p_intf);
+    if ( !p_sys->timer_listener )
+        goto error;
+
+
+
+    vlc_mutex_init (&p_sys->lock);
+    vlc_cond_init (&p_sys->wait);
+
+    if ( vlc_clone (&p_sys->thread, Run, p_intf, VLC_THREAD_PRIORITY_LOW) )
+        goto error;
+
+
+    return VLC_SUCCESS;
+
+    error:
+        if ( p_sys->playlist_listener )
+        {
+            vlc_playlist_Lock (playlist);
+            if ( p_sys->player_listener )
+                vlc_player_RemoveListener (player, p_sys->player_listener);
+            if ( p_sys->timer_listener )
+                vlc_player_RemoveTimer (player, p_sys->timer_listener);
+            vlc_playlist_RemoveListener (playlist, p_sys->playlist_listener);
+            vlc_playlist_Unlock (playlist);
+        }
+        free (p_sys);
+        return VLC_EGENERIC;
+}
+
+static void Close (vlc_object_t *p_this)
+{
+    intf_thread_t *p_intf = (intf_thread_t *) p_this;
+    intf_sys_t *p_sys = p_intf->p_sys;
+    vlc_playlist_t *playlist = p_sys->playlist;
+
+    vlc_cancel (p_sys->thread);
+    vlc_join (p_sys->thread, NULL);
+
+    vlc_vector_clear (&p_sys->queue);
+    vlc_UrlClean (&p_sys->p_submit_url);
+
+    vlc_playlist_Lock (playlist);
+    vlc_player_RemoveListener (vlc_playlist_GetPlayer (playlist), p_sys->player_listener);
+    vlc_playlist_RemoveListener (playlist, p_sys->playlist_listener);
+    vlc_playlist_Unlock (playlist);
+
+    free (p_sys);
+}
+
+static void *Run (void *data)
+{
+    intf_thread_t *p_intf = data;
+    int canc = vlc_savecancel ();
+    bool b_wait = 0;
+    char* request, *payload;
+
+    intf_sys_t *p_sys = p_intf->p_sys;
+
+    while ( 1 )
+    {
+        vlc_restorecancel (canc);
+
+        if ( b_wait )
+            vlc_tick_wait (vlc_tick_now () + VLC_TICK_FROM_SEC (60)); // wait for 1 min
+
+        vlc_mutex_lock (&p_sys->lock);
+        mutex_cleanup_push (&p_sys->lock) ;
+
+                while ( p_sys->queue.size == 0 )
+                    vlc_cond_wait (&p_sys->wait, &p_sys->lock);
+
+        vlc_cleanup_pop ();
+        vlc_mutex_unlock (&p_sys->lock);
+        canc = vlc_savecancel ();
+
+        payload = PreparePayload (p_intf);
```

Comment:
In order to avoid 2 consecutive lock/unlock from here and from PreparePayload.
You could Copy the Song to a local variable while holding the lock and pass it to PreparePayload() (that won't hold the lock anymore)

Reply:
PreparePayload need to be protected be the lock too (you should unlock after calling this funcction)